internal/handlers/common/update.go

```diff
@@ -236,6 +243,71 @@ func processPopFieldExpression(doc *types.Document, update *types.Document) (boo
 	return changed, nil
 }
 
+// processRenameFieldExpression changes document according to $rename operator.
+// If the document was changed it returns true.
+func processRenameFieldExpression(doc *types.Document, update *types.Document) (bool, error) {
+	renameExpression := update.SortFieldsByKey()
+
+	var changed bool
+
+	skipErrReg := regexp.MustCompile(`^types\.getByPath: types\.Document\.Get: key not found:.+`)
+
+	for _, key := range renameExpression.Keys() {
+		renameRawValue, err := renameExpression.Get(key)
+		if err != nil {
+			// if $rename field does not exist, don't change anything
+			continue
+		}
+
+		if key == "" || renameRawValue == "" {
+			return changed, NewWriteErrorMsg(ErrEmptyName, "An empty update path is not valid.")
+		}
+
+		sourcePath := types.NewPathFromString(key)
+
+		renameValue, ok := renameRawValue.(string)
+		if !ok {
+			return changed, NewWriteErrorMsg(
+				ErrBadValue,
+				fmt.Sprintf("the 'to' field for $rename must be a string: %s: %v", key, renameRawValue),
+			)
+		}
+
+		if key == renameRawValue {
+			return changed, NewWriteErrorMsg(
+				ErrBadValue,
+				fmt.Sprintf("The source and target field for $rename must differ: %s: %#v", key, renameValue),
+			)
+		}
+
+		targetPath := types.NewPathFromString(renameValue)
+
+		// Get value to move
+		val, err := doc.GetByPath(sourcePath)
+		if err != nil {
+			// TODO refactor `getByPath` to make this check more elegant
+			if skipErrReg.MatchString(err.Error()) {
+				continue
+			}
+
+			return changed, NewWriteErrorMsg(ErrUnsuitableValueType, err.Error())
```

Comment:
I beleive there was a test case which caused it to fail if we continue on any error so we handle `types.Document.Get: key not found` error specially? Could we create an error like ErrDocumentPathNotFound and handle it accordingly? Or would this cause too much change in codebase? :thinking:

Reply:
I was thinking to create some underlying error type like this one, the limitation I've seen though, is that creating new error variable in the same way as we do almost everywhere (`ErrDocumentPathNotFound = fmt.Errorf("...")`) limits our ability to add the invalid key inside the message, as we currently do and as MongoDB does. We could make some extended logic (like error codes), but that's more complex and is the reason why I left `TODO` here, so we can discuss the best approach :)

Reply:
If I understand correctly what's going on, we need to skip the error if the key was not found in the document. To do so we can return `KeyNotFound` error type from `types.Document`.`Get` and check for it here.

Reply:
done